File: src/execgo_playground/chaos/engine.py
from __future__ import annotations

import copy
import random
from typing import Any

from ..models import ChaosProfile, PlanStage, StandardPlan
from ..observability.timeline import TimelineRecorder
from ..utils import deep_set


class ChaosEngine:
    def __init__(self, harness, timeline: TimelineRecorder, run_id: str, framework: str, scenario_id: str) -> None:
        self.harness = harness
        self.timeline = timeline
        self.run_id = run_id
        self.framework = framework
        self.scenario_id = scenario_id
        self._fixture_dirty = False
        self._runtime_dirty = False

    def apply_plan_phase(self, plan: StandardPlan, profile: ChaosProfile) -> StandardPlan:
        if profile.target_phase != "plan" or not self._should_fire(profile):
            return plan
        mutated = copy.deepcopy(plan)
        for action in profile.actions:
            if action.kind == "inject_invalid_action" and mutated.stages and mutated.stages[0].task_graph.tasks:
                mutated.stages[0].task_graph.tasks[0].type = action.params.get("type", "ghost-tool")
            if action.kind == "inject_bad_dependency" and mutated.stages and mutated.stages[0].task_graph.tasks:
                mutated.stages[0].task_graph.tasks[0].depends_on.append(action.params.get("dependency", "missing-task"))
        self._record("plan", profile.id, "applied")
        return mutated

    def apply_submit_phase(self, stage: PlanStage, profile: ChaosProfile) -> PlanStage:
        if profile.target_phase != "submit" or not self._should_fire(profile):
            return stage
        mutated = copy.deepcopy(stage)
        for action in profile.actions:
            if action.kind == "drop_required_binding" and mutated.bindings:
                mutated.bindings.pop(0)
            if action.kind == "force_timeout_budget" and mutated.task_graph.tasks:
                mutated.task_graph.tasks[0].timeout = int(action.params.get("timeout_ms", 1))
        self._record("submit", profile.id, "applied", stage_id=stage.stage_id)
        return mutated

    def apply_runtime_phase(self, profile: ChaosProfile) -> None:
        if profile.target_phase not in {"runtime", "poll"} or not self._should_fire(profile):
            return
        for action in profile.actions:
            if action.kind == "restart_service":
                self.harness.restart_service(action.params["service"])
                self._runtime_dirty = True
            elif action.kind == "kill_service":
                self.harness.kill_service(action.params["service"])
                self._runtime_dirty = True
            elif action.kind == "resource_pressure":
                self.harness.update_service_resources(
                    action.params["service"],
                    cpus=action.params.get("cpus"),
                    memory=action.params.get("memory"),
                )
                self._runtime_dirty = True
            elif action.kind == "fixture_mode":
                self.harness.configure_fixture_mode(
                    latency_ms=int(action.params.get("latency_ms", 0)),
                    fail_mode=action.params.get("fail_mode", "none"),
                )
                self._fixture_dirty = True
        self._record(profile.target_phase, profile.id, "applied")

    def apply_verify_phase(self, profile: ChaosProfile, verification_payload: dict[str, Any]) -> dict[str, Any]:
        if profile.target_phase != "verify" or not self._should_fire(profile):
            return verification_payload
        mutated = copy.deepcopy(verification_payload)
        for action in profile.actions:
            if action.kind == "drop_verification_field":
                deep_set(mutated, action.params["path"], None)
        self._record("verify", profile.id, "applied")
        return mutated

    def cleanup(self) -> None:
        if self._fixture_dirty:
            self.harness.reset_fixture_mode()
        try:
            if self._runtime_dirty:
                self.harness.restart_service("runtime")
                self.harness.restart_service("execgo")
        except Exception:
            pass

    def _record(self, phase: str, profile_id: str, status: str, stage_id: str | None = None) -> None:
        self.timeline.record(
            run_id=self.run_id,
            framework=self.framework,
            scenario_id=self.scenario_id,
            phase=phase,
            stage_id=stage_id,
            event_type="chaos",
            status=status,
            metadata={"profile": profile_id},
        )

    def _should_fire(self, profile: ChaosProfile) -> bool:
        rng = random.Random(profile.seed or 0)
        return rng.random() <= profile.probability

File: src/execgo_playground/chaos/engine.ts
import { randomInt } from "node:crypto";
import { type ChaosProfile, type PlanStage, type StandardPlan, normalizePlanStage, normalizeStandardPlan } from "../models.js";
import { TimelineRecorder } from "../observability/timeline.js";
import { deepSet } from "../utils.js";

type HarnessLike = {
  restartService(service: string): void;
  killService(service: string): void;
  updateServiceResources(service: string, options: { cpus?: string; memory?: string }): void;
  configureFixtureMode(payload: { latency_ms?: number; fail_mode?: string }): Promise<void> | void;
  resetFixtureMode(): Promise<void> | void;
};

export class ChaosEngine {
  private fixtureDirty = false;
  private runtimeDirty = false;

  constructor(
    private readonly harness: HarnessLike,
    private readonly timeline: TimelineRecorder,
    private readonly runId: string,
    private readonly framework: string,
    private readonly scenarioId: string,
  ) {}

  applyPlanPhase(plan: StandardPlan, profile: ChaosProfile): StandardPlan {
    if (profile.target_phase !== "plan" || !this.shouldFire(profile)) {
      return plan;
    }
    const mutated = normalizeStandardPlan(structuredClone(plan));
    for (const action of profile.actions) {
      if (action.kind === "inject_invalid_action" && mutated.stages[0]?.task_graph.tasks[0]) {
        mutated.stages[0].task_graph.tasks[0].type = String(action.params.type ?? "ghost-tool");
      }
      if (action.kind === "inject_bad_dependency" && mutated.stages[0]?.task_graph.tasks[0]) {
        mutated.stages[0].task_graph.tasks[0].depends_on.push(String(action.params.dependency ?? "missing-task"));
      }
    }
    this.record("plan", profile.id, "applied");
    return mutated;
  }

  applySubmitPhase(stage: PlanStage, profile: ChaosProfile): PlanStage {
    if (profile.target_phase !== "submit" || !this.shouldFire(profile)) {
      return stage;
    }
    const mutated = normalizePlanStage(structuredClone(stage) as Record<string, unknown>);
    for (const action of profile.actions) {
      if (action.kind === "drop_required_binding" && mutated.bindings.length > 0) {
        mutated.bindings.shift();
      }
      if (action.kind === "force_timeout_budget" && mutated.task_graph.tasks[0]) {
        mutated.task_graph.tasks[0].timeout = Number(action.params.timeout_ms ?? 1);
      }
    }
    this.record("submit", profile.id, "applied", stage.stage_id);
    return mutated;
  }

  async applyRuntimePhase(profile: ChaosProfile): Promise<void> {
    if (!["runtime", "poll"].includes(profile.target_phase) || !this.shouldFire(profile)) {
      return;
    }
    for (const action of profile.actions) {
      if (action.kind === "restart_service") {
        this.harness.restartService(String(action.params.service));
        this.runtimeDirty = true;
      } else if (action.kind === "kill_service") {
        this.harness.killService(String(action.params.service));
        this.runtimeDirty = true;
      } else if (action.kind === "resource_pressure") {
        this.harness.updateServiceResources(String(action.params.service), {
          cpus: action.params.cpus === undefined ? undefined : String(action.params.cpus),
          memory: action.params.memory === undefined ? undefined : String(action.params.memory),
        });
        this.runtimeDirty = true;
      } else if (action.kind === "fixture_mode") {
        await this.harness.configureFixtureMode({
          latency_ms: Number(action.params.latency_ms ?? 0),
          fail_mode: String(action.params.fail_mode ?? "none"),
        });
        this.fixtureDirty = true;
      }
    }
    this.record(profile.target_phase, profile.id, "applied");
  }

  applyVerifyPhase(profile: ChaosProfile, verificationPayload: Record<string, unknown>): Record<string, unknown> {
    if (profile.target_phase !== "verify" || !this.shouldFire(profile)) {
      return verificationPayload;
    }
    const mutated = structuredClone(verificationPayload);
    for (const action of profile.actions) {
      if (action.kind === "drop_verification_field") {
        deepSet(mutated, String(action.params.path), null);
      }
    }
    this.record("verify", profile.id, "applied");
    return mutated;
  }

  async cleanup(): Promise<void> {
    if (this.fixtureDirty) {
      await this.harness.resetFixtureMode();
    }
    try {
      if (this.runtimeDirty) {
        this.harness.restartService("runtime");
        this.harness.restartService("execgo");
      }
    } catch {
      // Cleanup is best effort so the run result can still be inspected.
    }
  }

  private record(phase: string, profileId: string, status: string, stageId?: string): void {
    this.timeline.record({
      run_id: this.runId,
      framework: this.framework,
      scenario_id: this.scenarioId,
      phase,
      stage_id: stageId,
      event_type: "chaos",
      status,
      metadata: { profile: profileId },
    });
  }

  private shouldFire(profile: ChaosProfile): boolean {
    if (profile.probability <= 0) {
      return false;
    }
    if (profile.probability >= 1) {
      return true;
    }
    const seed = profile.seed || 0;
    const value = ((seed * 9301 + 49297 + randomInt(0, 1)) % 233280) / 233280;
    return value <= profile.probability;
  }
}
